### PR TITLE
WIP : Add KIT in process mode

### PR DESCRIPTION
**Pull request by Darshan-upadhyay1110**
Signed-off-by: Darshan-upadhyay1110 <darshan.upadhyay@collabora.com>

Change-Id: I153bdeed50e86c9a97a60591f2644eb6b4c51f00

 - have a mode KITINPROCESS (limited to --enable-debug) where everything is running as a single process
 - global function that will know if we're in kit-in-process mode,=> Util::isKitInProcess()

### CODE REVIEW

File: wsd/DocumentBroker.cpp [hunk 1]
```diff
@@ -1121,13 +1121,15 @@ bool DocumentBroker::download(const std::shared_ptr<ClientSession>& session, con
...
-                        return false;
+                    if (!Util::isKitInProcess()) {
+                        int process = Util::spawnProcess(command, args);
+                        int status = -1;
+                        const int rc = ::waitpid(process, &status, 0);
+                        if (rc != 0)
+                        {
+                            LOG_ERR("Conversion from " << extension << " to " << newExtension << " failed (" << rc << ").");
+                            return false;
+                        }
```
Comment (vmiklos) [inline wsd/DocumentBroker.cpp:1132]: Command used to be a separate process. It has a main function. Probably you need an else branch here, so that you invoke that main function on a thread?

File: configure.ac [hunk 2]
```diff
@@ -192,6 +192,10 @@ AC_ARG_ENABLE([androidapp],
                               to work similarly to the iOS app, from the JavaScript and the pseudo WebSocket
                               message plumbing point of view.]))
 
+AC_ARG_ENABLE([kitinprocess],
+              AS_HELP_STRING([--enable-kitinprocess],
```
Comment (vmiklos) [inline configure.ac:196]: Nit: probably this is not needed, this mode could be enabled at runtime with a cmdline switch and just be enabled in debug mode. The more ifdefs we have, the more likely CI doesn't catch a build breakage.

File: kit/ForKit.cpp [hunk 3]
```diff
@@ -383,50 +383,61 @@ static int createLibreOfficeKit(const std::string& childRoot,
     LOG_DBG("Forking a coolkit process with jailId: " << jailId << " as spare coolkit #"
                                                       << spareKitId << '.');
     const auto startForkingTime = std::chrono::steady_clock::now();
+    pid_t pid = 0;
+    if (Util::isKitInProcess()) {
```
Comment (mmeeks) [inline kit/ForKit.cpp:387]: Makes sense of course - we should of course guard against doing this twice - having two threads in lokit_main will be a nightmare =) can you add a LOG_ERR and perhaps a sleeping loop for that mode - I guess we have this pre-forking logic that will try to always launch a 2nd idle process when one is used so - sleeping rather than asserting makes sense for that.

**Review by mmeeks — COMMENTED** [review]
Looks like a fine start - I expect we'll want a different coolwsd process - this is most likely the easiest way to get everything linked into one binary.

Create a cool-onedoc binary or something in the Makefile.am - that links in all of the coolwsd and coolforkit source files - and then tweak / unwind the conflicts / overlaps etc.

Remove the 'spawnProcess' that starts the forkit - and instead run its 'main' or some other entry point you can add.

To avoid both 'main' functions conflicting - it's likely we'll need to split out two very small 'coolwsd-main' and 'forkit-main' files that just call an extern "C" int coolwsd_main(int argv, char **argv); type function each - since we can't have two main functions in the same linked object.

Does that help ?

File: common/Util.hpp [hunk 4]
```diff
@@ -1300,6 +1300,8 @@ int main(int argc, char**argv)
      */
     bool isFuzzing();
 
+    bool isKitInProcess();
```
Comment (hfiguiere) [inline common/Util.hpp:1303]: shouldn't this me `const` ?

Comment (Darshan-upadhyay1110) [reply common/Util.hpp:1303]: just a function declaration so i think it is okay :)

Comment (hfiguiere) [reply common/Util.hpp:1303]: duh I misread the scope. nevermind.

File: kit/ForKit.cpp (hunk 3 above)
Comment (Darshan-upadhyay1110) [reply kit/ForKit.cpp:387]: Question : i am not sure about which mode you are refering here. is it the KitInProcess where the new thread has been created and we call lokit_main ?

and the LOG_ERR is also for that main function call right ?

File: wsd/COOLWSD.cpp [hunk 5]
```diff
@@ -3724,22 +3699,23 @@ class PrisonerRequestDispatcher final : public WebSocketHandler
             LOG_TRC("Child connection with URI [" << COOLWSD::anonymizeUrl(request.getUrl())
                                                   << ']');
             Poco::URI requestURI(request.getUrl());
-#ifndef KIT_IN_PROCESS
-            if (requestURI.getPath() == FORKIT_URI)
-            {
-                if (socket->getPid() != COOLWSD::ForKitProcId)
+            if(!Util::isKitInProcess()) {
```
Comment (Ashod) [inline wsd/COOLWSD.cpp:3702]: Please use clang-format to have uniform formatting everywhere. You can configure vscode to apply formatting only on edited lines.

File: common/Util.cpp [hunk 6]
```diff
@@ -325,6 +325,13 @@ namespace Util
         return std::getenv("DISPLAY") != nullptr;
     }
 
+
+    bool isKitInProcess()
+    {
+        const char* kitInProcess = std::getenv("KIT_IN_PROCESS");
+        return (kitInProcess != nullptr && strcmp(kitInProcess, "true") == 0);
```
Comment (Ashod) [inline common/Util.cpp:332]: This can be static. Also, is there a reason to have it set to "true" and "false"? We can simply define the `KIT_IN_PROCESS` for 'true' and not define it for 'false'. And avoid this comparison (just check for `!= nullptr`).

File: kit/ForKit.cpp [hunk 7]
```diff
@@ -383,18 +380,28 @@ static int createLibreOfficeKit(const std::string& childRoot,
...
+        std::thread ([childRoot, jailId, sysTemplate, loTemplate, queryVersion]
+            {
+                Util::setThreadName("kit_in_process");
+                lokit_main(childRoot, jailId, sysTemplate, loTemplate, NoCapsForKit, NoSeccomp,
+                   queryVersion, DisplayVersion, spareKitId);
+            }).detach();
+        lokitProcessIsRunning = true;
+    }
+    else {
+        pid = fork();
```
Comment (Ashod) [inline kit/ForKit.cpp:395]: Please run clang-format, or apply proper formatting. This is all wrong.

File: kit/ForKit.cpp [hunk 8]
```diff
@@ -383,18 +380,28 @@ static int createLibreOfficeKit(const std::string& childRoot,
...
-        // Child
-
-        // Close the pipe from coolwsd
-        close(0);
+        std::thread ([childRoot, jailId, sysTemplate, loTemplate, queryVersion]
+            {
+                Util::setThreadName("kit_in_process");
+                lokit_main(childRoot, jailId, sysTemplate, loTemplate, NoCapsForKit, NoSeccomp,
+                   queryVersion, DisplayVersion, spareKitId);
+            }).detach();
```
Comment (Ashod) [inline kit/ForKit.cpp:391]: Why are we detaching? Why isn't the thread owned by coolwsd?

Can you share the design of this whole PR please?

File: kit/ForKit.cpp [hunk 9]
```diff
@@ -423,11 +430,11 @@ static int createLibreOfficeKit(const std::string& childRoot,
             LOG_INF("Forked kit [" << pid << ']');
             childJails[pid] = childRoot + jailId;
         }
-
-#ifndef KIT_IN_PROCESS
-        UnitKit::get().launchedKit(pid);
-#endif
+        if (!Util::isKitInProcess()) {
```
Comment (Ashod) [inline kit/ForKit.cpp:433]: Please do not use javascript formatting in C++ code. Apply clang-format if it's too much work.

File: wsd/debug-main.cpp [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+        Util::setThreadName("debug main kit_in_process");
+        int result = forkit_main(args.size(), argv);
+
+        // Fulfill the promise with the result.
+        resultPromise.set_value(result);
+    }).detach();
+
+    // Wait for the thread to finish and get the result using a future.
+    std::future<int> resultFuture = resultPromise.get_future();
+    int result = resultFuture.get();
```
Comment (Ashod) [inline wsd/debug-main.cpp:46]: The proper way of waiting for a thread to finish is to join the thread.

There is no good reason to detach the thread and use a promise/future.
```
int result = 0;
std::thread worker = std::thread ([&] {
        Util::setThreadName("debug main kit_in_process");
        result = forkit_main(args.size(), argv);
    });

worker.join();
return result;
```

File: wsd/COOLWSD-main.cpp [hunk 11]
```diff
@@ -0,0 +1,40 @@
+/*
```
Comment (Ashod) [inline wsd/COOLWSD-main.cpp:1]: Not clear to me why we need all these new files. Is there an explanation on the design somewhere?

File: kit/ForKit.cpp [hunk 12]
```diff
@@ -366,6 +361,8 @@ static void cleanupChildren()
     }
 }
 
+static bool lokitProcessIsRunning = false;
```
Comment (Ashod) [inline kit/ForKit.cpp:364]: Please add comments so it is clear what the various variables/functions do. It's not clear what this is supposed to represent and how it's supposed to work.

File: wsd/COOLWSD.cpp [hunk 13]
```diff
@@ -6383,7 +6356,9 @@ void forwardSigUsr2()
 int main(int argc, char** argv)
 {
     SigUtil::setUserSignals();
-    SigUtil::setFatalSignals("wsd " COOLWSD_VERSION " " COOLWSD_VERSION_HASH);
+    if (!Util::isKitInProcess()) {
+        SigUtil::setFatalSignals("wsd " COOLWSD_VERSION " " COOLWSD_VERSION_HASH);
+    }
```
Comment (Ashod) [inline wsd/COOLWSD.cpp:6361]: Why? Why remove the fatal signal in 'kit in process' mode?

**Review by Ashod — COMMENTED** [review]
Thanks @Darshan-upadhyay1110. I left some comments and suggestions, as well as questions.

File: wsd/COOLWSD-main.cpp (hunk 11 above)
Comment (eszkadev) [reply wsd/COOLWSD-main.cpp:1]: @Ashod comment above will explain: https://github.com/CollaboraOnline/online/pull/7329#pullrequestreview-1655289137

File: common/Util.cpp (hunk 6 above)
Comment (eszkadev) [reply common/Util.cpp:332]: @Darshan-upadhyay1110 as you have separate binaries now: regular one and  "single process" one, that env variable is not needed I think.
You should make this true when you link single process code probably.

File: wsd/COOLWSD.cpp (hunk 13 above)
Comment (Darshan-upadhyay1110) [reply wsd/COOLWSD.cpp]: ohh yes, sorry we only need to remove fatal signal for Forkit and Kit not for COOLWSD ( pushing that change...)

File: kit/ForKit.cpp (hunk 7 above)
Comment (Darshan-upadhyay1110) [reply kit/ForKit.cpp]: thanks @Ashod i updated the formatting mistakes but there might be some still present sorry for that. i am still working on this patch so will update all formatting mistakes  soon...

**Comment by JaumePujantell** [comment]
Push my current local state (not working) so @caolanm  can see it.

**Comment by mmeeks** [comment]
Eventually we merged this in another pull request; thanks Darshan =)
